File: javascript/linked_list/sorted_insert.js

const DoublyLinkedList = class {
    constructor() {
        this.head = null;
        this.tail = null;
    }

    insertNode(nodeData) {
        let node = new DoublyLinkedListNode(nodeData);

        if (this.head == null) {
            this.head = node;
        } else {
            this.tail.next = node;
            node.prev = this.tail;
        }

        this.tail = node;
    }
};

function printDoublyLinkedList(node, sep, ws) {
    while (node != null) {
        ws.write(String(node.data));

        node = node.next;

        if (node != null) {
            ws.write(sep);
        }
    }
}

function sortedInsert(head, data) {
    // create our doubly linked list node 
    const newNode = new DoublyLinkedListNode();
    newNode.data = data;
​
    // base case 1: our current linked list node is null
    if (head === null) {
        // just return our newNode 
        return newNode;
​
    // base case 2: data <= our current linked list node's data
    } else if (data <= head.data) {
        newNode.next = head;
        head.prev = newNode;
        // return our newNode in this case since 
        // it's the new head of our linked list 
        return newNode;
    } else {
        // recursively call this function on the next node
        // of our linked list 
        // this will continue to re-curse until we find 
        // the spot where the data belongs 
        const rest = sortedInsert(head.next, data);
        // connect rest with our head 
        head.next = rest;
        rest.prev = head;
        // and we're done!
        // return our head node 
        return head;
    }
}

File: python/linked_list/sorted_insert.py
class DoublyLinkedListNode:
    def __init__(self, node_data):
        self.data = node_data
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def insert_node(self, node_data):
        node = DoublyLinkedListNode(node_data)

        if not self.head:
            self.head = node
        else:
            self.tail.next = node
            node.prev = self.tail


        self.tail = node

def print_doubly_linked_list(node, sep, fptr):
    while node:
        fptr.write(str(node.data))

        node = node.next

        if node:
            fptr.write(sep)

# Complete the sortedInsert function below.

#
# For your reference:
#
# DoublyLinkedListNode:
#     int data
#     DoublyLinkedListNode next
#     DoublyLinkedListNode prev
#
#
def sortedInsert(head, data):
    new_node = DoublyLinkedListNode(data)
    if not head:
        return new_node
    current = head
    if current.data > data:
        new_node.next = current
        current.prev = new_node
        return new_node
    while current.next:
        if data > current.data and data <= current.next.data:
            new_node.next = current.next
            current.next.prev = new_node
            current.next=new_node
            new_node.prev=current
            break
        else:
            current = current.next
    else:
        current.next=new_node
        new_node.prev=current
    return head